### PR TITLE
Openshift cloud provider part .1

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
This is the first pr to add cloud provider support to openshift clusters.
It mainly consists of refactoring and adding the config parameters.

I'm opening this as it already grew a lot and having it even bigger would make reviewing really painful.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mrIncompetent** [comment]
/hold 

https://github.com/kubermatic/kubermatic/pull/3036 needs to be merged before as this PR is based on it.

File: api/pkg/controller/monitoring/monitoring_controller.go [hunk 1]
```diff
@@ -305,6 +305,7 @@ func (c *Controller) sync(key string) error {
 	if clusterFromCache.Status.NamespaceName == "" {
 		glog.V(6).Infof("skipping cluster %s because it has no namespace yet", key)
 		c.enqueueAfter(clusterFromCache, healthCheckPeriod)
+		return nil
```
Comment (mrIncompetent) [inline api/pkg/controller/monitoring/monitoring_controller.go:308]: That triggered a lot of errors in the monitoring controller

File: api/pkg/test/helper.go [hunk 2]
```diff
@@ -8,8 +8,12 @@ import (
 	"github.com/pmezard/go-difflib/difflib"
 )
 
-func CompareOutput(t *testing.T, name, output string, update bool) {
-	golden, err := filepath.Abs(filepath.Join("testdata", name+".golden"))
+func CompareOutput(t *testing.T, name, output string, update bool, suffix string) {
```
Comment (mrIncompetent) [inline api/pkg/test/helper.go:11]: I added the `suffix` param so the gold files can have a specific suffix.
This enables code highlighting + autocomplete on most IDE's - makes those golden files much nicer to read

**Review by nikhita — APPROVED** [review]
Sorry for the delay, overall lgtm :tada: 

/lgtm
/approve

/assign @kron4eg 
for the monitoring part

**Pull request merged**